src/object/cli_obj.c

```diff
@@ -5117,12 +5142,21 @@ obj_ec_valid_shard_get(struct obj_auxi_args *obj_auxi, uint8_t *tgt_bitmap,
 	struct dc_object	*obj = obj_auxi->obj;
 	uint32_t		grp_start = grp_idx * obj_get_grp_size(obj);
 	uint32_t		shard_idx = grp_start + *tgt_idx;
-	bool			force_ec_degrade = false;
+	bool			force_ec_degrade = obj_auxi->force_degraded;
 	bool			ec_degrade = false;
 	int			rc = 0;
 
-	if (need_retry_redundancy(obj_auxi) || obj_auxi->force_degraded)
-		force_ec_degrade = true;
+	if (need_retry_redundancy(obj_auxi)) {
+		struct obj_auxi_tgt_list *failed_list;
+		uint32_t tgt_id;
+
+		D_ASSERT(obj_auxi->failed_tgt_list != NULL);
+		failed_list = obj_auxi->failed_tgt_list;
+		tgt_id = obj->cob_shards->do_shards[shard_idx].do_target_id;
+
+		if (tgt_in_failed_tgts_list(tgt_id, failed_list))
+			force_ec_degrade = true;
```

Comment:
OK, i see what this PR try to fix.
But the logic of "nvme_io_err" flag seems still not fully correct, the below code when exit this func looks not right to me -
	if (ec_degrade) {
		obj_auxi->tx_uncertain = 0;
		obj_auxi->csum_retry = 0;
		obj_auxi->nvme_io_err = 0;
	}

why clear those flags right after the first degraded shard fetch?
For example if we have EC4P2G1 obj, get -DER_NVME_IO from both shard 0 and shard 1's fetch, when retry the fetch, after force degraded fetch for shard 0, the next shard 1 fetch also need to force degraded right? but the obj_auxi->nvme_io_err was cleared when force degraded for shard 0, then shard 1 still will get -DER_NVME_IO after retry.

I think we need to fix it if I am not wrong, but could be in a following patch.
BTW, those flags and retry logic really confuse, we may need to refine it later at appropriate time.

Reply:
yeah, the retry logic needs be refined. To be honest, I didn't fully follow the client retry code when I introduce the nvme io error retry, I just followed the csum error retry code.

Reply:
>why clear those flags right after the first degraded shard fetch?
For example if we have EC4P2G1 obj, get -DER_NVME_IO from both shard 0 and shard 1's fetch, when retry the fetch, after force degraded fetch for shard 0, the next shard 1 fetch also need to force degraded right? but the obj_auxi->nvme_io_err was cleared when force degraded for shard 0, then shard 1 still will get -DER_NVME_IO after retry.

Hmm, let me update the patch, make this clearer. thanks